Tests/SentryTests/Integrations/SentryNetworkTrackerTests.swift

```diff
@@ -263,6 +257,96 @@ class SentryNetworkTrackerTests: XCTestCase {
         assertStatus(status: .undefined, state: .completed, response: URLResponse())
     }
     
+    func testResumeAfterCompleted_OnlyOneSpanCreated() {
+        let task = createDataTask()
+        let sut = fixture.getSut()
+        let transaction = startTransaction()
+        
+        sut.urlSessionTaskResume(task)
+        task.state = .completed
+        sut.urlSessionTaskResume(task)
+
+        assertOneSpanCreated(transaction)
+    }
+    
+    func testResumeAfterCancelled_OnlyOneSpanCreated() {
+        let task = createDataTask()
+        let sut = fixture.getSut()
+        let transaction = startTransaction()
+        
+        sut.urlSessionTaskResume(task)
+        task.state = .canceling
+        sut.urlSessionTaskResume(task)
+
+        assertOneSpanCreated(transaction)
+    }
+    
+    // Altough we only run this test above the below specified versions, we exped the
+    // implementation to be thread safe
+    @available(tvOS 10.0, *)
+    @available(OSX 10.12, *)
+    @available(iOS 10.0, *)
+    func testResumeCalledMultipleTimesConcurrent_OneSpanCreated() {
+        let task = createDataTask()
+        let sut = fixture.getSut()
+        let transaction = startTransaction()
+        
+        let queue = DispatchQueue(label: "SentryNetworkTrackerTests", qos: .userInteractive, attributes: [.concurrent, .initiallyInactive])
+        let group = DispatchGroup()
+        
+        // The number is kept small for the CI to not take to long.
+        // If you really want to test this increase to 1_000_000 or so.
+        for _ in 0...10_000 {
+            group.enter()
+            queue.async {
+                sut.urlSessionTaskResume(task)
+                task.state = .completed
+                group.leave()
+            }
+        }
+        
+        queue.activate()
+        group.waitWithTimeout(timeout: 500)
+        
+        assertOneSpanCreated(transaction)
+    }
+    
+    // Altough we only run this test above the below specified versions, we exped the
+    // implementation to be thread safe
+    @available(tvOS 10.0, *)
+    @available(OSX 10.12, *)
+    @available(iOS 10.0, *)
+    func testChangeStateMultipleTimesConcurrent_OneSpanFinished() {
+        let task = createDataTask()
+        let sut = fixture.getSut()
+        let transaction = startTransaction()
+        sut.urlSessionTaskResume(task)
+        
+        let queue = DispatchQueue(label: "SentryNetworkTrackerTests", qos: .userInteractive, attributes: [.concurrent, .initiallyInactive])
+        let group = DispatchGroup()
+        
+        // The number is kept small for the CI to not take to long.
+        // If you really want to test this increase to 1_000_000 or so.
+        for _ in 0...100_000 {
+            group.enter()
+            queue.async {
+                task.state = .completed
+                group.leave()
+            }
+        }
+        
+        queue.activate()
+        group.waitWithTimeout(timeout: 500)
```

Comment:
How often does this test break if you remove the fix? Say you run it 10 times, how many break?

These tests that attempt to verify race conditions often add up in terms of slowing the build for little value (unless they were proven to reproduce enough times).

Ideally we'd use synchronization mechanisms with the code being tested to have a reproducible code path that would validate your assumptions. 

Here for example there's no `thread.sleep` or `setTimeout` kind of strategy. Only 'timeout' arguments are to avoid a deadlock blocking CI forever so it would timeout and fail the build. 

It tests concurrent code by synchronizing steps along the execution flow:

https://github.com/getsentry/sentry-dotnet/blob/28ec35d278b2fa15d356af41079ab81a956fe364/test/Sentry.Tests/Internals/SentryScopeManagerTests.cs#L222-L282

Same here testing the background transport. Testing if the max queue items is reached, events start getting dropped without actually slowing down the build or being flaky:

https://github.com/getsentry/sentry-dotnet/blob/a9304a0a4b4702d0e62e2703d55c66483d27a0e5/test/Sentry.Tests/Internals/BackgroundWorkerTests.cs#L194-L222

Reply:
The tests are high-speed. I just lowered the timeout to 100 ms. This doesn't mean we wait for 100 ms, but we wait a maximum of 100 ms. If it doesn't complete until then, the test fails. I added this, so we avoid hanging tests. I agree with what you stated above. Ideally, we would also add tests as in .NET, but I think the tests here also add value. They helped me to uncover issues I didn't know exist before. I need to rethink our whole concurrent testing strategy to achieve better tests as in .NET, which takes some time.